### PR TITLE
rgw: extending existing ssl support for vault KMS

### PR DESCRIPTION
**Pull request by thotz**
Currently RGW can communicate with vault using SSL with certain limitations. Libcurl uses system certs and it is enabled by generic option `rgw_verify_ssl` which is to enable SSL with any http server and RGW
Introducing two options for RGW to extend existing support for SSL in Vault KMS
`rgw_crypt_vault_verify_ssl` -> exclusive option to enable SSL for vault 
`rgw_crypt_vault_ssl_cert = <file path for ca cert>`

Fixes: https://tracker.ceph.com/issues/47776
Signed-off-by: Jiffin Tony Thottan <jthottan@redhat.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_http_client.cc [hunk 1]
```diff
@@ -565,6 +565,9 @@ int RGWHTTPClient::init_request(rgw_http_req_data *_req_data)
     curl_easy_setopt(easy_handle, CURLOPT_SSL_VERIFYPEER, 0L);
     curl_easy_setopt(easy_handle, CURLOPT_SSL_VERIFYHOST, 0L);
     dout(20) << "ssl verification is set to off" << dendl;
+  } else if (!ca_path.empty()) {
+    curl_easy_setopt(easy_handle, CURLOPT_CAINFO, ca_path);
+    dout(20) << "using customed ca cert for ssl" << dendl;
```
Comment (mattbenjamin) [inline src/rgw/rgw_http_client.cc:570]: custom?

**Review by mattbenjamin — COMMENTED** [review]
this looks very clean to me

**Comment by mattbenjamin** [comment]
@thotz did the docs get updated?

**Comment by thotz** [comment]
> @thotz did the docs get updated?

Yeah I need to update the doc

File: src/common/legacy_config_opts.h [hunk 2]
```diff
@@ -1509,6 +1509,8 @@ OPTION(rgw_crypt_vault_addr, OPT_STR) // Vault server base address
 OPTION(rgw_crypt_vault_prefix, OPT_STR) // Optional URL prefix to Vault secret path
 OPTION(rgw_crypt_vault_secret_engine, OPT_STR) // kv, transit or other supported secret engines
 OPTION(rgw_crypt_vault_namespace, OPT_STR) // Vault Namespace (only availabe in Vault Enterprise Version)
+OPTION(rgw_crypt_vault_verify_ssl, OPT_BOOL) // should we try to verify vault's ssl
```
Comment (leseb) [inline src/common/legacy_config_opts.h:1512]: It would be nice to support client cert and client private key as well.

Comment (thotz) [reply src/common/legacy_config_opts.h:1512]: @leseb : The second commit have option for vault client [ca cert](https://github.com/ceph/ceph/pull/37730/commits/491997b66b6c29407947d2ea923c44f174e8a45f). That will be enough for the tls authentication , Am I missing something?

**Comment by thotz** [comment]
> @thotz did the docs get updated?

@mattbenjamin : Yup

**Comment by mattbenjamin** [comment]
@mdw-at-linuxbox please review

File: src/common/legacy_config_opts.h (hunk 2 above)
Comment (leseb) [reply src/common/legacy_config_opts.h:1508]: Ok for the cert, what about the private key?

Comment (thotz) [reply src/common/legacy_config_opts.h:1508]: Either I can use separate option for the private key using [CURLOPT_SSH_PRIVATE_KEYFILE](https://curl.se/libcurl/c/CURLOPT_SSH_PRIVATE_KEYFILE.html) or we can point to directory which contains bundle [CURLOPT_CAPATH](https://curl.se/libcurl/c/CURLOPT_CAPATH.html)

Comment (thotz) [reply src/common/legacy_config_opts.h:1508]: @leseb any suggestions?

Comment (mdw-at-linuxbox) [reply src/common/legacy_config_opts.h:1508]: The "verify_ssl" option should default to true.  The documentation should make it clear that setting it "false" is not recommended, dangerous, and nearly always avoidable.

The client key/cert stuff shoud be part of a separate PR - and it has to include TTL / periodic renewal logic to be useful.  Note that vault agent can already do that, so this feature may not be necessary.

Comment (thotz) [reply src/common/legacy_config_opts.h:1506]: I have updated PR with suggested changes and client key/cert will address as separate PR

File: src/common/options.cc [hunk 3]
```diff
@@ -7059,6 +7059,10 @@ std::vector<Option> get_rgw_options() {
     .set_default(true)
     .set_description("Should RGW verify the vault server SSL certificate."),
 
+    Option("rgw_crypt_vault_ssl_cert", Option::TYPE_STR, Option::LEVEL_ADVANCED)
+    .set_default("")
+    .set_description("Path for customed ca certificate for accessing vault server"),
```
Comment (mattbenjamin) [inline src/common/options.cc:7064]: "custom" not "customed"

Comment (mattbenjamin) [reply src/common/options.cc:7064]: Also, maybe just say "Path to CA certificate for accessing Vault server"--if we provide a CA cert, it's just a requirement of the setup, maybe not anything actually special or custom.

**Review by mattbenjamin — APPROVED** [review]
looks good to me--hopefully marcus will approve

**Comment by thotz** [comment]
> @thotz please rebase. @mdw-at-linuxbox please take another look.

PTAL @mdw-at-linuxbox

**Comment by TRYTOBE8TME** [comment]
I've tested this against the s3-tests suite and it's running good.

**Pull request merged**